### PR TITLE
ENG-11130 recover using shutdown snapshot

### PR DESCRIPTION
**Pull request by vtkstef**
[pro pull request](https://github.com/VoltDB/pro/pull/1535)

### CODE REVIEW

File: src/frontend/org/voltdb/RealVoltDB.java [hunk 1]
```diff
@@ -2373,6 +2387,7 @@ public Integer get() {
                 .nodeStateSupplier(m_statusTracker.getNodeStateSupplier())
                 .addAllowed(m_config.m_enableAdd)
                 .safeMode(m_config.m_safeMode)
+                .terminusNonce(getTerminusNonce())
```
Comment (akhanzode) [inline src/frontend/org/voltdb/RealVoltDB.java]: Do you want to do this for old verbs as well? I would think it should be for "start" only.

File: src/frontend/org/voltdb/RestoreAgent.java [hunk 2]
```diff
@@ -685,11 +694,19 @@ else if (snaptxnId < cmdlogmap.get(cmdpart)) {
...
         }
-        LOG.debug("Gathered " + snapshotInfos.size() + " snapshot information");
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Gathered " + snapshotInfos.size() + " snapshot information");
+        }
         sendLocalRestoreInformation(maxLastSeenTxn, snapshotInfos);
 
         // Negotiate with other hosts about which snapshot to restore
         SnapshotInfo infoWithMinHostId = getRestorePlan();
+        if (infoWithMinHostId != null && infoWithMinHostId.nonce.equals(m_terminusNonce)) {
```
Comment (akhanzode) [inline src/frontend/org/voltdb/RestoreAgent.java]: infoWithMinHostId.nonce.trim()? Do you mean !equal here?

File: src/frontend/org/voltdb/probe/HostCriteria.java [hunk 3]
```diff
@@ -102,15 +105,19 @@ public HostCriteria(JSONObject jo) {
...
     public HostCriteria(boolean paused, UUID configHash, UUID meshHash,
             boolean enterprise, StartAction startAction, boolean bare,
-            int hostCount, NodeState nodeState, boolean addAllowed, boolean safeMode) {
+            int hostCount, NodeState nodeState, boolean addAllowed,
+            boolean safeMode, String terminusNonce) {
         checkArgument(configHash != null, "config hash is null");
         checkArgument(meshHash != null, "mesh hash is null");
         checkArgument(startAction != null, "start action is null");
         checkArgument(hostCount > 0, "host count %s is less then one", hostCount);
+        checkArgument(terminusNonce == null || !terminusNonce.trim().isEmpty(),
```
Comment (akhanzode) [inline src/frontend/org/voltdb/probe/HostCriteria.java]: !isEmpty() thought you are checking for empty.

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 4]
```diff
@@ -219,6 +222,8 @@ protected MeshProber(NavigableSet<String> coordinators,
         checkArgument(coordinators.size() <= hostCountSupplier.get(),
                 "host count %s is less then the number of coordinators %s",
                 hostCountSupplier.get(), coordinators.size());
+        checkArgument(terminusNonce == null || !terminusNonce.trim().isEmpty(),
```
Comment (akhanzode) [inline src/frontend/org/voltdb/probe/MeshProber.java]: !isEmpty() ??

File: src/frontend/org/voltdb/probe/MeshProber.java [hunk 5]
```diff
@@ -504,6 +516,7 @@ private void determineStartActionIfNecessary(Map<Integer, HostCriteria> hostCrit
         int bare = 0; // node has no recoverable artifacts (Command Logs, Snapshots)
         int unmeshed = 0;
         int operational = 0;
+        int haveTerminus = 0;
```
Comment (akhanzode) [inline src/frontend/org/voltdb/probe/MeshProber.java]: haveTerminus not required as you are checking size()

File: tests/frontend/org/voltdb/regressionsuites/TestShutdownSave.java [hunk 6]
```diff
@@ -136,7 +137,28 @@ public boolean accept(File f) {
...
+        m_config.startUp(false);
+        client2 = this.getClient();
+
+        ClientResponse cr = client2.callProcedure("@Statistics", "PROCEDUREPROFILE", 0);
+        assertEquals("statistics invocation failed: " + cr.getStatusString() , ClientResponse.SUCCESS, cr.getStatus());
+        VoltTable t = cr.getResults()[0];
+        while (t.advanceRow()) {
+            assertNotSame("detected replayed transactions", "KV.insert", t.getString(1));
+        }
+
```
Comment (akhanzode) [inline tests/frontend/org/voltdb/regressionsuites/TestShutdownSave.java]: Dont you want to check if data is there after restore?

File: src/frontend/org/voltdb/probe/MeshProber.java (hunk 5 above)
Comment (vtkstef) [inline src/frontend/org/voltdb/probe/MeshProber.java]: I am checking that there are enough terminuses there to set it in the determination. Basically nodes - kfactor must have terminus for pober to return a determination with a terminus

File: tests/frontend/org/voltdb/regressionsuites/TestShutdownSave.java (hunk 6 above)
Comment (vtkstef) [inline tests/frontend/org/voltdb/regressionsuites/TestShutdownSave.java]: ok

**Pull request merged**